floqast_sftp/app.py

```diff
@@ -0,0 +1,269 @@
+    return client
+
+
+def get_file_name(_period):
+    """
+    Construct a file name for the given period.
+
+    Parameters
+    ----------
+    _period: str
+        ISO-8601 formatted string specifying a date (YYYY-MM-DD).
+
+    Returns
+    -------
+    str
+        File name.
+    """
+    # use full month name
+    _date = date.fromisoformat(_period)
+    _period_month = _date.strftime("%B-%Y")
+
+    # Include creation timestamp in file name
+    _timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
+
+    name = f"Sage-Balances-{_period_month}-{_timestamp}.csv"
+    return name
+
+
+def get_period_count(event):
+    """
+    Parse the period count from the EventBridge event dictionary.
+
+    Parameters
+    ----------
+    event: dict
+        Event dictionary passed to lambda_handler
+
+    Returns
+    -------
+    int
+        Number of periods requested.
+
+    """
+
+    if "period_count" not in event:
+        raise ValueError("'period_count' not provided")
+
+    try:
+        period_count = int(event["period_count"])
+    except (TypeError, ValueError) as exc:
+        raise ValueError("'period_count' must be an integer.") from exc
+
+    if period_count < 1:
+        raise ValueError("'period_count' must be greater than 0")
+
+    return period_count
+
+
+def get_previous_month(_date):
+    """
+    Calculate the date one month previous to the given date.
+
+    Parameters
+    ----------
+    _date: datetime.date
+        Date to calculate previous month for.
+
+    Returns
+    -------
+    datetime.date
+        Date one month previous to the given date.
+
+    """
+    if _date.month == 1:
+        _year = _date.year - 1
+        return _date.replace(year=_year, month=12)
+
+    _month = _date.month - 1
+    return _date.replace(month=_month)
+
+
+def get_balances_csv(when):
+    """
+    Fetch the balances CSV from lambda-mips-api for the given activity period.
+
+    Parameters
+    ----------
+    when: datetime.date
+        Date to pass to lambda-mips-api for calculating the balance activity period.
+
+    Returns
+    -------
+    (str, file)
+        Tuple of file name and a file-like object.
+
+    """
+    url = _csv_url + f"&target_date={when}"
+
+    response = requests.get(url, stream=True)
+    response.raise_for_status()
+    file_obj = io.StringIO(response.text)
+
+    # read target period from CSV
+    csv_dict = csv.DictReader(file_obj)
+    row = next(csv_dict)
+    period = row["PeriodStart"]
+    # and reset file position
+    file_obj.seek(0)
+
+    filename = get_file_name(period)
+
+    return filename, file_obj
+
+
+def put_sftp_file(client, name, file_obj):
+    """
+    Wrapper around SFTPClient.putfo
+
+    Parameters
+    ----------
+    client: paramiko.SFTPClient
+        SFTP client.
+
+    name: str
+        Remote file name
+
+    file_obj
+        File-like object to upload.
+
+    Returns
+    -------
+    None
+
+    """
+    client.putfo(fl=file_obj, remotepath=name)
```

Comment:
you should make sure to close the connection by wrapping this in try finally block. take a look  https://stackoverflow.com/a/8293225/1094247

Reply:
thanks, great catch! I wasn't closing the session anywhere. I do want to re-use the session for all of the uploads (each activity period has it's own CSV file) so I put the try/finally block around the main lambda_handler loop.